### PR TITLE
Improved command

### PR DESCRIPTION
**Pull request by zMoooooritz**
I wanted to add some more abilities for customization to widget_command. Since the current implementation is fairly restrictive.

Most of the work in `Update` was for no reason repeated every iteration so I moved most of it into the "constructor" (could be done for other widgets as well).
This means I can save the final correct typed values (font, color, ...) directly in CommandWidget and save time in `Update`.

~~`config.go` got therefore some new values and we may think about adding the `Rectangle` part in widget_command to `config.go` as well since this could be used for other widgets as well (widget_time)?~~

Edit: I have moved the layout logic into a separate file since it can be used in at least two widgets so far. Currently the "layout" is supported for `widget_command` and `widget_time`.

For now the implementation allows by intention frames that are not fully inside the button because `DrawString` does not allow for text alignment. The currently used frame format is `[posX]x[posY]+[width]x[height]`.

What are your thoughts on all these changes?

Off topic: Do you also have artifacts when using colored widgets. For example using the time widget with the color `#ff0000` does result in red areas apart from the text itself.

### CODE REVIEW

**Comment by muesli** [comment]
> What are your thoughts on all these changes?

Will give them a review asap!

> Off topic: Do you also have artifacts when using colored widgets. For example using the time widget with the color `#ff0000` does result in red areas apart from the text itself.

No, I don't see any artifacts. Which StreamDeck model do you have? Can you try to take a photo? Does it only happen for specific colors?

**Comment by zMoooooritz** [comment]
> > Off topic: Do you also have artifacts when using colored widgets. For example using the time widget with the color `#ff0000` does result in red areas apart from the text itself.
> 
> No, I don't see any artifacts. Which StreamDeck model do you have? Can you try to take a photo? Does it only happen for specific colors?

Fixed! Looks way better now!

File: widget.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func NewBaseWidget(base string, index uint8, action, actionHold *ActionConfig, b
 }
 
 // NewWidget initializes a widget.
-func NewWidget(base string, kc KeyConfig, bg image.Image) (Widget, error) {
+func NewWidget(dev *streamdeck.Device, base string, kc KeyConfig, bg image.Image) (Widget, error) {
 	bw := NewBaseWidget(base, kc.Index, kc.Action, kc.ActionHold, bg)
```
Comment (muesli) [inline widget.go:93]: Maybe instead of passing `dev` to every single widget, we could initialize `BaseWidget` with it, and make it accessible that way.

**Comment by muesli** [comment]
Looks great, love the layout helper! Note the few linter warnings!

File: layouts.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	for i := 0; i < frameCount; i++ {
+		if len(frameReps) < i+1 {
+			frame := l.defaultFrame(frameCount, i)
+			l.frames = append(l.frames, frame)
+			continue
+		}
+
+		frame, err := formatFrame(frameReps[i])
+		if err != nil {
+			fmt.Println("using default frame: ", err)
```
Comment (muesli) [inline layouts.go:56]: Something we should comment out before merging.

**Comment by zMoooooritz** [comment]
> Maybe instead of passing `dev` to every single widget, we could initialize `BaseWidget` with it, and make it accessible that way.

Yes good one! This also allows us to drop the `dev` argument from each `Update` in the widgets.

> Looks great, love the layout helper! Note the few linter warnings!

Thank you! Mhhh I am confused :confused:, my local GoLint did not catch these mistakes.

> Something we should comment out before merging.

Are you sure? This message only pops up if a layout is faulty meaning the format is wrong, in case no layout (or only a parital layout) is defined this will be caught in the first `if` and this message will not be displayed.
In case you still want to comment it out go ahead :)

I did also adjust the README but it is becoming more and more a mess, mostly because of redundancy and the array (";" seperated values).
Maybe we could describe in one paragraph which optional settings exist (i.e. font, color, layout, flatten, ...)
And have a matrix like table that shows which (optional) settings can be used for which widget.
I case you agree with these ideas I could create a seperate PR?

**Comment by muesli** [comment]
> Thank you! Mhhh I am confused confused, my local GoLint did not catch these mistakes.

Make sure to install `golangci-lint`. It's a fantastic meta-linter that combines a whole variety of different linters & checks. We have a config file in the repository for it (`.golangci.yml`), so the output you see locally and on the CI should be the same.

File: layouts.go (hunk 2 above)
Comment (zMoooooritz) [reply layouts.go:56]: See my penultimate post

Comment (muesli) [reply layouts.go:56]: Makes sense, thank you for clarifying!

**Comment by muesli** [comment]
Sorry this took a bit longer to review, but due to the size of this PR I just had to find the time for it first :smile:

Thank you for another great contribution!

**Comment by zMoooooritz** [comment]
No problem, thank you for merging it and maintaining this project in such a conscientious way :+1:

**Pull request merged**